From: Paul Pluzhnikov <ppluzhnikov@google.com>
Date: Sun, 18 Oct 2015 09:39:25 -0700
Subject: Fix PR binutils/19147 -- off by one when printing NT_FILE note.

[ Upstream 595712bb ]

---
diff --git a/binutils/readelf.c b/binutils/readelf.c
--- a/binutils/readelf.c
+++ b/binutils/readelf.c
@@ -14411,7 +14411,7 @@ print_core_note (Elf_Internal_Note *pnote)
 	  (int) (4 + 2 * addr_size), _("End"),
 	  (int) (4 + 2 * addr_size), _("Page Offset"));
   filenames = descdata + count * 3 * addr_size;
-  while (--count > 0)
+  while (count-- > 0)
     {
       bfd_vma start, end, file_ofs;
 
-- 
