From dc0a93c756035fb319f45b53585e9fbb9f8e36fa
From: Peter S. Mazinger <ps.m@gmx.net>
Date: Tue, 12 Apr 2011 20:01:42 +0000
Subject: add lcong48[_r]

Signed-off-by: Peter S. Mazinger <ps.m@gmx.net>
Signed-off-by: Bernhard Reutner-Fischer <rep.dot.nop@gmail.com>
---
diff --git a/libc/stdlib/Makefile.in b/libc/stdlib/Makefile.in
--- a/libc/stdlib/Makefile.in
+++ b/libc/stdlib/Makefile.in
@@ -15,7 +15,7 @@ CSRC-y := \
 	abort.c getenv.c mkdtemp.c realpath.c canonicalize.c mkstemp.c \
 	rand.c random.c random_r.c setenv.c div.c ldiv.c lldiv.c \
 	getpt.c drand48-iter.c jrand48.c \
-	jrand48_r.c lrand48.c lrand48_r.c mrand48.c mrand48_r.c nrand48.c \
+	jrand48_r.c lcong48.c lrand48.c lrand48_r.c mrand48.c mrand48_r.c nrand48.c \
 	nrand48_r.c rand_r.c srand48.c srand48_r.c seed48.c seed48_r.c \
 	valloc.c a64l.c l64a.c __uc_malloc.c
 CSRC-$(UCLIBC_HAS_ADVANCED_REALTIME) += posix_memalign.c
diff --git a/libc/stdlib/lcong48.c b/libc/stdlib/lcong48.c
new file mode 100644
--- /dev/null
+++ b/libc/stdlib/lcong48.c
@@ -0,0 +1,29 @@
+/* vi: set sw=4 ts=4: */
+/* Licensed under the LGPL v2.1, see the file COPYING.LIB in this tarball. */
+
+#include <features.h>
+
+#if defined __USE_SVID || defined __USE_XOPEN
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+
+static int __lcong48_r (unsigned short int param[7], struct drand48_data *buffer)
+{
+	/* Store the given values. */
+	memcpy (buffer->__x, &param[0], sizeof (buffer->__x));
+	buffer->__a = ((uint64_t) param[5] << 32 | (uint32_t) param[4] << 16 | param[3]);
+	buffer->__c = param[6];
+	buffer->__init = 1;
+
+	return 0;
+}
+# ifdef __USE_MISC
+strong_alias(__lcong48_r,lcong48_r)
+# endif
+
+void lcong48 (unsigned short int param[7])
+{
+	(void) __lcong48_r (param, &__libc_drand48_data);
+}
+#endif
--
cgit v0.9.1
